### PR TITLE
feat: Add custom scripts and hook based setup

### PR DESCRIPTION
**Pull request by morten-olsen**
Hey love the project.

Had an idea to make adding custom stuff easier without having to worry about merge conflicts, and without having to reset custom scripts when updating.

So the idea is that the user can add a `lua/custom` folder, which will be loaded on startup before anything else in NvChad gets loaded. Inside this the user can create files which uses a "hook" system for running different tasks at different points in the startup.

This is one of the first times i write `lua` so the code is not as pretty as could be but it should illustrate the purpose.

I have for illustrative purposes added a custom module which loads the Tmux plugin using the code from https://github.com/NvChad/NvChad/wiki/Chad-user-configs#tmux-integration-in-vim-buffers--now-in-lua- rewritten using hooks

**File changes:**
* **lua/core/custom.lua**: Checkes if a custom module exists and loads it
* **lua/core/hooks.lua**: The hook module for invoking and registering hooks
* **lua/core/init.lua**: Added loading of `core.custom` as well as added a hook that runs after NvChad is done setting up
* **lua/core/mappings.lua**: Added a hook that runs after core mapping has been set up
* **lua/custom/init.lua**: Demo custom module
* **lua/custom/tmux.lua**: Example of introducing tmux as a custom module using hooks
* **lua/plugins/init.lua**: Added a hook that runs after the plugin list gets loaded

### CODE REVIEW

**Comment by siduck76** [comment]
sounds cool , But I dont think we should be including a tmux config ( tmux.lua ) since its use specific or just put it as an example but everything commented.

**Comment by dmitry12** [comment]
Nice, I like the approach!

> sounds cool , But I dont think we should be including a tmux config ( tmux.lua ) since its use specific or just put it as an example but everything commented.

Agreed, maybe having it commented out in `custom/init.lua` will work both for people using tmux and as an example for custom mappings.

**Comment by siduck76** [comment]
also load    "core.custom" at last in core_modules so : 

```
"core.custom" 
"core.options",
"core.autocmds",
"core.mappings",
   ```
   
  this would prevent custom.lua's stuffs getting overwritten by other core files

**Comment by G-Rowell** [comment]
I really like having an example,
maybe instead of keeping `tmux.lua` in the repo, we could post it on our website? https://nvchad.netlify.app/

Otherwise we just have an example file hanging around, which is a lot more than just a commented out example line.

---

My other comment would be that it deviates from having a centrally located `plugins/init.lua` to Packer use all the plugins, but I'm okay with this as it's specifically for user customisation.


---

Thanks so much for the awesome work @morten-olsen!!!!


G

**Comment by G-Rowell** [comment]
Again, my advice here is just opinion, so I'm open to being wrong.

I do think that we could consider re-structuring a little.
As generally we want the user to **only** modify `chadrc.lua` (as modifications outside of this file may cause notable issues).

Maybe we move *all files that the user is allowed to change to `/lua/custom/`**?
This would include `chadrc.lua`?

**Comment by phatsk** [comment]
I agree with @siduck76 that this should be at the bottom of the core modules - I was worried that some other thing would overwrite custom things I had set.

Also @morten-olsen in your `tmux.lua`, how are you registering the settings like `tmux..copy_sync.enble`? Are those configurations already part of that plugin? And what calls `tmux_mappings`? I guess an example of how to define configurations/mappings for _any_ plugin would be helpful here.

**Comment by G-Rowell** [comment]
I agree an example would be useful, but for such a complex (but needed!) thing, we should have detailed examples on our docs instead. 
Keep the code as code, and docs as docs I think.

https://nvchad.netlify.app/


And @phatsk, mappings for each custom file is called by this line https://github.com/NvChad/NvChad/pull/368/files#diff-cd2cb877ba027cebfab5d54935a09e407b07b9108ddad1fbc019e6a49710554fR117
Which is a hook.

**Comment by G-Rowell** [comment]
I'd be happy to write up a page on our docs for this?
That way we actively have an example in our docs before this is merged?

**Comment by G-Rowell** [comment]
@morten-olsen 

Hey hey Morten!
What can I do to help get this finished? I've got a bit of time tonight and would love to lend a hand.

**Comment by G-Rowell** [comment]
That being said, looks pretty close to being finished tbh

**Comment by morten-olsen** [comment]
Think so as Well, or at least in a stage that would add value if merged. Think the only remaining question is if `lua/custom/init.lua` should be removed from the repo an instead be explained in the readme or the wiki?

**Comment by G-Rowell** [comment]
I think remove it, it can always be added later.

Any custom file that we have present will just give us temptation to modify it, potentially causing merge conflicts for users.

@siduck76 @Akianonymus @ashincoder We should consider merging this into `NvChad/test` now

**Review by siduck76 — APPROVED** [review]

File: lua/custom/tmux.lua [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        }
+      })
+    end
+  }
+end)
+
+hooks.add("setup_mappings", function(map)
+  local opt = {}
+
+ local function tmux_mappings()
```
[the remark this answers is not in the document]
Comment (morten-olsen) [reply lua/custom/tmux.lua]: Yes, removed the function wrapping in the version in the comments

**Pull request merged**